File: java/solitaire/game/src/main/java/org/silnith/game/search/WorkerThreadDepthFirstSearch.java
package org.silnith.game.search;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.BlockingDeque;
import java.util.concurrent.ConcurrentLinkedQueue;
import java.util.concurrent.LinkedBlockingDeque;
import java.util.concurrent.TimeUnit;

import org.silnith.game.Game;
import org.silnith.game.GameState;
import org.silnith.game.move.Move;
import org.silnith.game.util.LinkedNode;

/**
 * A game tree search algorithm implementation that performs a depth-first
 * search running in parallel across multiple threads.
 * 
 * @param <M> the move type for the game
 * @param <B> the board type for the game
 */
public class WorkerThreadDepthFirstSearch<M extends Move<B>, B>
        extends GameTreeSearcher<M, B> {

	private final List<Worker> workers;
	private final BlockingDeque<LinkedNode<GameState<M, B>>> queue;
	private final Collection<List<GameState<M, B>>> wins;
	private volatile boolean cancelled;

	/**
	 * Initializes a depth-first search across the specified number of threads.
	 * 
	 * @param game the game
	 * @param initialState the initial game state from which to begin the search
	 * @param numThreads the number of threads to use for searching the game tree
	 */
	public WorkerThreadDepthFirstSearch(final Game<M, B> game, final GameState<M, B> initialState,
			final int numThreads) {
		super(game);
		if (initialState == null) {
			throw new IllegalArgumentException("Initial state cannot be null.");
		}
		this.workers = new ArrayList<>(numThreads);
		this.queue = new LinkedBlockingDeque<>();
		this.wins = new ConcurrentLinkedQueue<>();
		this.cancelled = false;

		this.queue.addLast(new LinkedNode<>(initialState));

		for (int i = 0; i < numThreads; i++) {
		    this.workers.add(new Worker());
		}
	}

	@Override
    public void close() {
	    cancelled = true;
    }

	@Override
    public Collection<List<GameState<M, B>>> call() throws InterruptedException {
        final ThreadGroup threadGroup = new ThreadGroup("SearcherWorkers");
        final List<Thread> threads = new ArrayList<>(workers.size());
        for (final Worker worker : workers) {
            final Thread thread = new Thread(threadGroup, worker, "Worker-" + threads.size());
            threads.add(thread);
            thread.start();
        }
        
        boolean finished;
        do {
            Thread.sleep(TimeUnit.SECONDS.toMillis(1));
            
            finished = queue.isEmpty() && !isWorking();
        } while (!finished && !cancelled);
        
        cancelled = true;
        
        for (final Thread thread : threads) {
            thread.join();
        }
        
        return Collections.unmodifiableCollection(wins);
    }

    @Override
    protected int getQueueSize() {
        return queue.size();
    }

    @Override
    protected int getWinCount() {
        return wins.size();
    }

    @Override
    protected void queueNode(final LinkedNode<GameState<M, B>> node) {
        queue.addLast(node);
    }

    @Override
    protected void addWin(final List<GameState<M, B>> node) {
        wins.add(node);
    }

    private boolean isWorking() {
        for (final Worker worker : workers) {
            if (worker.isWorking()) {
                return true;
            }
        }
        return false;
    }

	private class Worker implements Runnable {

	    private volatile boolean working = false;

	    public boolean isWorking() {
            return working;
        }

		@Override
		public void run() {
		    try {
	            LinkedNode<GameState<M, B>> node = queue.pollLast(1, TimeUnit.SECONDS);
	            while (node != null && !cancelled) {
	                working = true;
	                examineNode(node);
	                working = false;

	                node = queue.pollLast(1, TimeUnit.SECONDS);
	            }
		    } catch (final InterruptedException e) {
                e.printStackTrace();
            }
		}

	}

}

File: csharp/Solitaire/Game/Search/WorkerThreadDepthFirstSearch.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Linq;
using System.Threading;
using System.Threading.Tasks;

namespace Silnith.Game.Search
{
    /// <summary>
    /// A game tree search algorithm implementation that performs a depth-first
    /// search running in parallel across multiple threads.
    /// </summary>
    /// <typeparam name="M">The move type for the game.</typeparam>
    /// <typeparam name="B">The board type for the game.</typeparam>
    public class WorkerThreadDepthFirstSearch<M, B> : GameTreeSearcher<M, B> where M : IMove<B>
    {
        private readonly int numWorkers;
        private readonly BlockingCollection<LinkedNode<GameState<M, B>>> stack;
        private readonly ConcurrentQueue<IReadOnlyList<GameState<M, B>>> wins;

        /// <summary>
        /// Constructs a depth-first search across the specified number of threads.
        /// </summary>
        /// <param name="game">The game.</param>
        /// <param name="initialGameState">The initial game state from which to begin the search.</param>
        /// <param name="numThreads">The number of threads to use for searching the game tree.</param>
        public WorkerThreadDepthFirstSearch(IGame<M, B> game, GameState<M, B> initialGameState, int numThreads)
            : base(game)
        {
            numWorkers = numThreads;
            stack = new BlockingCollection<LinkedNode<GameState<M, B>>>(new ConcurrentStack<LinkedNode<GameState<M, B>>>());
            wins = new ConcurrentQueue<IReadOnlyList<GameState<M, B>>>();

            stack.Add(new LinkedNode<GameState<M, B>>(initialGameState));
        }

        /// <inheritdoc/>
        protected override int QueueSize => stack.Count;

        /// <inheritdoc/>
        protected override int WinCount => wins.Count;

        /// <inheritdoc/>
        public override IEnumerable<IReadOnlyList<GameState<M, B>>> Search()
        {
            return SearchAsync(CancellationToken.None).GetAwaiter().GetResult();
        }

        /// <inheritdoc/>
        public override async Task<IEnumerable<IReadOnlyList<GameState<M, B>>>> SearchAsync(CancellationToken cancellationToken)
        {
            async Task DoWorkAsync()
            {
                await WorkAsync(cancellationToken);
            }

            List<Task> tasks = new List<Task>();
            for (int i = 0; i < numWorkers; i++)
            {
                tasks.Add(Task.Run(DoWorkAsync, cancellationToken));
            }

            await Task.WhenAll(tasks);

            return wins;
        }

        private Task WorkAsync(CancellationToken cancellationToken = default)
        {
            int millis = (int) TimeSpan.FromSeconds(1).TotalMilliseconds;
            while (stack.TryTake(out LinkedNode<GameState<M, B>> node, millis, cancellationToken))
            {
                ExamineNode(node);
            }
            return Task.CompletedTask;
        }

        /// <inheritdoc/>
        protected override void AddWin(IReadOnlyList<GameState<M, B>> node)
        {
            wins.Enqueue(node);
        }

        /// <inheritdoc/>
        protected override void QueueNode(LinkedNode<GameState<M, B>> node)
        {
            stack.Add(node);
        }
    }
}
